File: src/main/java/it/polimi/tiw/util/StringUtil.java
package it.polimi.tiw.util;

import java.util.regex.Pattern;

/**
 * Utility class providing methods for string validation and manipulation.
 */
public class StringUtil {

    /**
     * Regular expression pattern for validating email addresses.
     * The email must contain letters, numbers, and optionally symbols (+, _, ., -),
     * followed by an '@' symbol, a domain name, and a top-level domain between 2 and 6 characters.
     */
    private static final String EMAIL_REGEX = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,6}$";

    /**
     * Compiled pattern for email validation based on the EMAIL_REGEX.
     */
    private static final Pattern EMAIL_PATTERN = Pattern.compile(EMAIL_REGEX);

    /**
     * Regular expression pattern for validating usernames.
     * Usernames must be 1 to 32 characters long, containing lowercase letters, numbers, underscores, or periods.
     * The username cannot start or end with a period, nor have consecutive periods.
     */
    private static final String USERNAME_REGEX = "^(?!\\.)(?!.*\\.\\.)[a-z0-9._]{1,32}(?<!\\.)$";

    /**
     * Compiled pattern for username validation based on the USERNAME_REGEX.
     */
    private static final Pattern USERNAME_PATTERN = Pattern.compile(USERNAME_REGEX);

    /**
     * Regular expression pattern for validating passwords.
     * Passwords must be at least 8 characters long.
     */
    private static final String PASSWORD_REGEX = "^.{8,}$";

    /**
     * Compiled pattern for password validation based on the PASSWORD_REGEX.
     */
    private static final Pattern PASSWORD_PATTERN = Pattern.compile(PASSWORD_REGEX);

    /**
     * Checks if a string is null or empty.
     * @param string the string to check
     * @return true if the string is null or empty, false otherwise
     */
    public static boolean isNullOrEmpty(String string) {
        return string == null || string.isEmpty();
    }

    /**
     * Checks if a string contains spaces.
     * @param string the string to check
     * @return true if the string contains spaces, false otherwise
     */
    public static boolean hasSpaces(String string) {
        return string.contains(" ");
    }

    /**
     * Validates if a string's length falls within the specified range.
     * @param string the string to validate
     * @param minLength the minimum acceptable length
     * @param maxLength the maximum acceptable length
     * @return true if the string length is within the range, false otherwise
     */
    public static boolean isValidLength(String string, int minLength, int maxLength) {
        return (!isNullOrEmpty(string) && string.length() >= minLength && string.length() <= maxLength);
    }

    /**
     * Validates if the provided email has a valid format.
     * An email is considered valid if it is not null, not empty, does not contain spaces,
     * and matches the specified email pattern.
     * @param email the email address to validate
     * @return true if the email is valid, false otherwise
     */
    public static boolean isValidEmail(String email) {
        return (!isNullOrEmpty(email) && !hasSpaces(email) && EMAIL_PATTERN.matcher(email).matches() && isValidLength(email, 6, 64));
    }

    /**
     * Validates if the provided username has a valid format.
     * A username is considered valid if it is not null, not empty, does not contain spaces,
     * and matches the specified username pattern.
     * @param username the username to validate
     * @return true if the username is valid, false otherwise
     */
    public static boolean isValidUsername(String username) {
        return (!isNullOrEmpty(username) && !hasSpaces(username) && USERNAME_PATTERN.matcher(username).matches() && isValidLength(username, 1, 32));
    }

    /**
     * Validates if the provided password has a valid format.
     * A password is considered valid if it is not null, not empty, does not contain spaces,
     * and matches the specified password pattern.
     * @param password the password to validate
     * @return true if the password is valid, false otherwise
     */
    public static boolean isValidPassword(String password) {
        return (!isNullOrEmpty(password) && !hasSpaces(password) && PASSWORD_PATTERN.matcher(password).matches() && isValidLength(password, 8, 128));
    }

    /**
     * Validates if the provided album title is valid.
     * An album title is considered valid if it is not null and does not start with the '@' character.
     * @param title the album title to validate
     * @return true if the album title is valid, false otherwise
     */
    public static boolean isValidTitle(String title) {
        return (!isNullOrEmpty(title) && !title.startsWith("@") && isValidLength(title, 1, 64));
    }

    /**
     * Validates if the provided text is valid.
     * A text is considered valid if it is not null and its length is within the acceptable range.
     * @param text the text to validate
     * @return true if the text is valid, false otherwise
     */
    public static boolean isValidText(String text) {
        return (!isNullOrEmpty(text) && isValidLength(text, 1, 512));
    }

}

File: src/main/webapp/javascript/StringUtil.js
/**
 * Utility class providing methods for string validation and manipulation.
 * This class includes validation methods for emails, usernames, passwords,
 * titles, and generic texts, ensuring they meet specific format and length requirements.
 */
class StringUtil {

    /**
     * Regular expression for email validation.
     * Matches emails containing letters, numbers, and optional symbols (+, _, ., -),
     * followed by an '@' symbol, a domain name, and a top-level domain between 2 and 6 characters.
     */
    static EMAIL_REGEX = /^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,6}$/;

    /**
     * Regular expression for username validation.
     * Matches usernames 1 to 32 characters long, containing lowercase letters, numbers, underscores, or periods.
     * Usernames cannot start or end with a period, nor have consecutive periods.
     */
    static USERNAME_REGEX = /^(?!\.)(?!.*\.\.)[a-z0-9._]{1,32}(?<!\.)$/;

    /**
     * Regular expression for password validation.
     * Matches passwords with a minimum length of 8 characters.
     */
    static PASSWORD_REGEX = /^.{8,}$/;

    /**
     * Regular expression for title validation.
     * Matches titles that do not start with '@' and do not contain '<' or '>'.
     */
    static TITLE_REGEX = /^(?!@)[^<>]*$/;

    /**
     * Regular expression for text validation.
     * Matches texts that do not contain the characters '<' or '>'.
     */
    static TEXT_REGEX = /^[^<>]*$/;

    /**
     * Checks if a string is null or empty.
     * @param {string} string - The string to check.
     * @returns {boolean} - True if the string is null or empty, false otherwise.
     */
    static isNullOrEmpty(string) {
        return string == null || string.trim().length === 0;
    }

    /**
     * Checks if a string contains spaces.
     * @param {string} string - The string to check.
     * @returns {boolean} - True if the string contains spaces, false otherwise.
     */
    static hasSpaces(string) {
        return string.includes(' ');
    }

    /**
     * Validates if a string's length falls within the specified range.
     * @param {string} string - The string to validate.
     * @param {number} minLength - The minimum acceptable length.
     * @param {number} maxLength - The maximum acceptable length.
     * @returns {boolean} - True if the string length is within the range, false otherwise.
     */
    static isValidLength(string, minLength, maxLength) {
        return !this.isNullOrEmpty(string) && string.length >= minLength && string.length <= maxLength;
    }

    /**
     * Validates if the provided email has a valid format.
     * @param {string} email - The email address to validate.
     * @returns {boolean} - True if the email is valid, false otherwise.
     */
    static isValidEmail(email) {
        return (
            !this.isNullOrEmpty(email) &&
            !this.hasSpaces(email) &&
            this.EMAIL_REGEX.test(email) &&
            this.isValidLength(email, 6, 64)
        );
    }

    /**
     * Validates if the provided username has a valid format.
     * @param {string} username - The username to validate.
     * @returns {boolean} - True if the username is valid, false otherwise.
     */
    static isValidUsername(username) {
        return (
            !this.isNullOrEmpty(username) &&
            !this.hasSpaces(username) &&
            this.USERNAME_REGEX.test(username) &&
            this.isValidLength(username, 1, 32)
        );
    }

    /**
     * Validates if the provided password has a valid format.
     * @param {string} password - The password to validate.
     * @returns {boolean} - True if the password is valid, false otherwise.
     */
    static isValidPassword(password) {
        return (
            !this.isNullOrEmpty(password) &&
            !this.hasSpaces(password) &&
            this.PASSWORD_REGEX.test(password) &&
            this.isValidLength(password, 8, 128)
        );
    }

    /**
     * Validates if the provided album title is valid.
     * @param {string} title - The album title to validate.
     * @returns {boolean} - True if the album title is valid, false otherwise.
     */
    static isValidTitle(title) {
        return (
            !this.isNullOrEmpty(title) &&
            this.TITLE_REGEX.test(title) &&
            this.isValidLength(title, 1, 64)
        );
    }

    /**
     * Validates if the provided text is valid.
     * @param {string} text - The text to validate.
     * @returns {boolean} - True if the text is valid, false otherwise.
     */
    static isValidText(text) {
        return (
            !this.isNullOrEmpty(text) &&
            this.TEXT_REGEX.test(text) &&
            this.isValidLength(text, 1, 512)
        );
    }

}

// Export the class for use in other files
export default StringUtil;